e2e/crd_validation_test.go

```diff
@@ -0,0 +1,826 @@
+
+	// Apply GMP CRDs
+	applyCRDsOutput, err := exec.CommandContext(ctx, "kubectl", "--kubeconfig", kubeconfigPath, "apply", "-f", "../manifests/setup.yaml").CombinedOutput()
+	if err != nil {
+		t.Fatalf("%s\b%v", applyCRDsOutput, err)
+	}
+	t.Logf("%s\n", applyCRDsOutput)
+
+	// Wait for CRDs to be created - there seems to be race condition without this wait
+	if _, err := exec.CommandContext(ctx, "kubectl", "--kubeconfig", kubeconfigPath, "wait", "customresourcedefinition.apiextensions.k8s.io/clusternodemonitorings.monitoring.googleapis.com", "customresourcedefinition.apiextensions.k8s.io/clusterpodmonitorings.monitoring.googleapis.com", "customresourcedefinition.apiextensions.k8s.io/clusterrules.monitoring.googleapis.com", "customresourcedefinition.apiextensions.k8s.io/globalrules.monitoring.googleapis.com", "customresourcedefinition.apiextensions.k8s.io/operatorconfigs.monitoring.googleapis.com", "customresourcedefinition.apiextensions.k8s.io/podmonitorings.monitoring.googleapis.com", "customresourcedefinition.apiextensions.k8s.io/rules.monitoring.googleapis.com", "--for=create").CombinedOutput(); err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	// Load the test cluster kubeconfig
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	// Create a client for the test cluster
+	c, err := newKubeClient(config)
+	if err != nil {
+		t.Error(err)
+	}
+	return c
+}
+
+func cleanupKindCluster(t *testing.T, clusterName string) func() {
+	return func() {
+		out, err := exec.Command("kind", "delete", "cluster", "--name", clusterName).CombinedOutput()
+		t.Logf("%s\n", out)
+		if err != nil {
+			t.Log(err.Error())
+		}
+	}
+}
+
+func TestCRDValidation(t *testing.T) {
+	ctx := context.Background()
+
+	t.Parallel()
+
+	c := createKindCluster(t)
+
+	type test struct {
+		obj     client.Object
+		wantErr bool
+	}
+
+	run := func(t *testing.T, tests map[string]test) {
+		for name, tc := range tests {
+			t.Run(name, func(t *testing.T) {
+				err := c.Create(ctx, tc.obj)
+				switch {
+				case err == nil && !tc.wantErr:
+					// OK
+				case err != nil && !tc.wantErr:
+					t.Errorf("Unexpected error: %v", err)
+				case err == nil && tc.wantErr:
+					t.Errorf("Want error, but got none")
+				case err != nil && tc.wantErr:
+					t.Log(err)
+					// OK
+				}
+			})
+		}
+	}
+	t.Run("ClusterNodeMonitoring", func(t *testing.T) {
+		tests := map[string]test{
+			// ClusterNodeMonitoring tests
+			"scrape interval missing": {
+				obj: &monitoringv1.ClusterNodeMonitoring{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "scrape-interval-missing",
+						Namespace: "default",
+					},
+					Spec: monitoringv1.ClusterNodeMonitoringSpec{
+						Endpoints: []monitoringv1.ScrapeNodeEndpoint{
+							{},
+						},
+					},
+				},
+			},
+			"scrape interval malformed": {
+				obj: &monitoringv1.ClusterNodeMonitoring{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "scrape-interval-malformed",
```

Comment:
nit: Doing test case list instead of map and using Object.Meta.Name might be better (more deterministic given map random iteration and less things to type/read)